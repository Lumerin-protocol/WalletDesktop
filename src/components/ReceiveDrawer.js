import { CopyIcon, BaseBtn, Drawer, Flex } from './common'
import * as selectors from '../selectors'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import QRCode from 'qrcode.react'
import React from 'react'
import Web3 from 'web3'
const { clipboard } = window.require('electron')

const Body = styled.div`
  padding: 3.2rem 1.6rem;

  @media (min-height: 700px) {
    padding: 6.4rem 1.6rem;
  }
`

const Title = styled.div`
  line-height: 2rem;
  font-size: 1.6rem;
  font-weight: 600;
  text-shadow: 0 1px 1px ${p => p.theme.colors.darkShade};
`

const Address = styled.div`
  margin-top: 1.6rem;
  padding: 0.8rem 1.6rem;
  border-radius: 4px;
  background-color: ${p => p.theme.colors.lightShade};
  line-height: 1.6rem;
  font-size: 1.3rem;
  letter-spacing: 0.5px;
  font-weight: 600;
  text-shadow: 0 1px 1px ${p => p.theme.colors.darkShade};
  letter-spacing: normal;
`

const CopyBtn = styled(BaseBtn)`
  margin-top: 2.4rem;
  border: 0;
  padding: 0;
  font: inherit;
  width: 6.4rem;
  height: 6.4rem;
  border-radius: 3.2rem;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(126, 97, 248, 0.4);
  &:hover {
    background-color: rgba(126, 97, 248, 1);
  }
`

const BtnLabel = styled.div`
  opacity: ${p => (p.isCopied ? '1' : '0.5')};
  margin-top: 0.8rem;
  line-height: 1.6rem;
  font-size: 1.3rem;
  font-weight: 600;
  text-shadow: 0 2px 0 ${p => p.theme.colors.darkShade};
`

const Footer = styled.div`
  background-image: linear-gradient(to bottom, #272727, #323232);
  padding: 3.2rem 2.4rem;
  flex-grow: 1;
  height: 100%;
  align-items: center;
  justify-content: center;
  display: flex;

  @media (min-height: 700px) {
    padding: 6.4rem 2.4rem;
  }
`

const QRContainer = styled.div`
  background: white;
  padding: 1.6rem;

  & canvas {
    display: block;
  }
`

const QRmsg = styled.div`
  text-align: center;
  margin-bottom: -4rem;
  margin-top: 2.4rem;
  line-height: 1.6rem;
  font-size: 1.3rem;
  font-weight: 600;
  letter-spacing: 0.5px;
  text-shadow: 0 1px 1px ${p => p.theme.colors.darkShade};
`

class ReceiveDrawer extends React.Component {
  static propTypes = {
    onRequestClose: PropTypes.func.isRequired,
    address: PropTypes.string.isRequired,
    isOpen: PropTypes.bool.isRequired
  }

  state = {
    copyStatus: 'init'
  }

  onCopyToClipboardClick = () => {
    clipboard.writeText(this.props.address)
    this.setState({ copyStatus: 'copied' })
  }

  render() {
    const { onRequestClose, address, isOpen } = this.props
    const { copyStatus } = this.state

    return (
      <Drawer
        onRequestClose={onRequestClose}
        data-testid="receive-drawer"
        isOpen={isOpen}
        title="Receive Transaction"
      >
        <Flex.Column grow="1">
          <Body>
            <Flex.Column align="center">
              <Title>Your address</Title>
              <Address data-testid="address">{address}</Address>
              <CopyBtn
                data-testid="copy-btn"
                onClick={this.onCopyToClipboardClick}
              >
                <CopyIcon />
              </CopyBtn>
              <BtnLabel
                data-testid="btn-label"
                isCopied={copyStatus !== 'init'}
              >
                {copyStatus === 'init' ? 'Copy' : 'Copied to clipboard!'}
              </BtnLabel>
            </Flex.Column>
          </Body>
          <Footer>
            <QRContainer>
              <QRCode value={address} />
              <QRmsg>Scan address</QRmsg>
            </QRContainer>
          </Footer>
        </Flex.Column>
      </Drawer>
    )
  }
}

const mapStateToProps = state => {
  const address = selectors.getActiveWalletAddresses(state)[0]
  return {
    address: address ? Web3.utils.toChecksumAddress(address) : address
  }
}

export default connect(mapStateToProps)(ReceiveDrawer)
