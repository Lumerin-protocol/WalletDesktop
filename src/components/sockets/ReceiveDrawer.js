import withReceiveDrawerState from 'lumerin-wallet-ui-logic/src/hocs/withReceiveDrawerState';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import QRCode from 'qrcode.react';
import React from 'react';

import { BaseBtn, Drawer, Flex } from '../common';
import CopyIcon from '../icons/CopyIcon';

const Body = styled.div`
  padding: 3.2rem 1.6rem;

  @media (min-height: 700px) {
    padding: 6.4rem 1.6rem;
  }
`;

const Title = styled.div`
  line-height: 2rem;
  font-size: 1.6rem;
  font-weight: 600;
  text-shadow: 0 1px 1px ${p => p.theme.colors.darkShade};
  cursor: default;
`;

const Address = styled.div`
  margin-top: 1.6rem;
  padding: 0.8rem 1.6rem;
  border-radius: 4px;
  background-color: ${p => p.theme.colors.lightShade};
  line-height: 1.6rem;
  font-size: 1.3rem;
  letter-spacing: 0.5px;
  font-weight: 600;
  text-shadow: 0 1px 1px ${p => p.theme.colors.darkShade};
  letter-spacing: normal;
`;

const CopyBtn = styled(BaseBtn)`
  margin-top: 2.4rem;
  border: 0;
  padding: 0;
  font: inherit;
  width: 6.4rem;
  height: 6.4rem;
  border-radius: 3.2rem;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(126, 97, 248, 0.4);
  &:hover {
    background-color: rgba(126, 97, 248, 1);
  }
`;

const BtnLabel = styled.div`
  opacity: ${p => (p.isCopied ? '1' : '0.5')};
  margin-top: 0.8rem;
  line-height: 1.6rem;
  font-size: 1.3rem;
  font-weight: 600;
  text-shadow: 0 2px 0 ${p => p.theme.colors.darkShade};
`;

const Footer = styled.div`
  background-image: linear-gradient(to bottom, #272727, #323232);
  padding: 3.2rem 2.4rem;
  flex-grow: 1;
  height: 100%;
  align-items: center;
  justify-content: center;
  display: flex;

  @media (min-height: 700px) {
    padding: 6.4rem 2.4rem;
  }
`;

const QRContainer = styled.div`
  background: white;
  padding: 1.6rem;

  & canvas {
    display: block;
  }
`;

const QRmsg = styled.div`
  text-align: center;
  margin-bottom: -4rem;
  margin-top: 2.4rem;
  line-height: 1.6rem;
  font-size: 1.3rem;
  font-weight: 600;
  letter-spacing: 0.5px;
  text-shadow: 0 1px 1px ${p => p.theme.colors.darkShade};
`;

class ReceiveDrawer extends React.Component {
  static propTypes = {
    copyToClipboard: PropTypes.func.isRequired,
    onRequestClose: PropTypes.func.isRequired,
    copyBtnLabel: PropTypes.string.isRequired,
    address: PropTypes.string.isRequired,
    isOpen: PropTypes.bool.isRequired
  };

  render() {
    return (
      <Drawer
        onRequestClose={this.props.onRequestClose}
        data-testid="receive-drawer"
        isOpen={this.props.isOpen}
        title="Receive Transaction"
      >
        <Flex.Column grow="1">
          <Body>
            <Flex.Column align="center">
              <Title>Your address</Title>
              <Address data-testid="address">{this.props.address}</Address>
              <CopyBtn
                data-testid="copy-btn"
                onClick={this.props.copyToClipboard}
              >
                <CopyIcon />
              </CopyBtn>
              <BtnLabel
                data-testid="btn-label"
                isCopied={this.props.copyBtnLabel !== 'Copy to clipboard'}
              >
                {this.props.copyBtnLabel}
              </BtnLabel>
            </Flex.Column>
          </Body>
          <Footer>
            <QRContainer>
              <QRCode value={this.props.address} />
              <QRmsg>Scan address</QRmsg>
            </QRContainer>
          </Footer>
        </Flex.Column>
      </Drawer>
    );
  }
}

export default withReceiveDrawerState(ReceiveDrawer);
