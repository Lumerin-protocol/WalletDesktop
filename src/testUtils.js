/* eslint-disable require-path-exists/exists */
/* eslint-disable import/no-unresolved */
import { Provider as ClientProvider } from 'lumerin-wallet-ui-logic/src/hocs/clientContext';
import { Provider, createStore } from 'lumerin-wallet-ui-logic/src/store';
import { render, Simulate } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import { MemoryRouter } from 'react-router';
import createClient from './client';
import { merge } from 'lodash';
import theme from 'lumerin-wallet-ui-logic/src/theme';
import React from 'react';

import config from '../config';

/*
 * The same render method of 'react-testing-library' but wrapped with
 * Redux store Provider and 'styled-components' ThemeProvider.
 *
 * element      : the React element you want to wrap
 * initialState : optional state to initialize the Redux store
 *
 * Returns the same object as render() but with an additional 'store'
 * property useful for dispatching actions inside your tests.
 */
export function reduxRender(element, initialState) {
  const client = createClient(config, createStore, initialState);

  const renderResult = render(
    <ClientProvider value={client}>
      <Provider store={client.store}>
        <ThemeProvider theme={theme}>
          <React.Fragment>{element}</React.Fragment>
        </ThemeProvider>
      </Provider>
    </ClientProvider>
  );
  return { ...renderResult, store: client.store };
}

/*
 * The same as reduxRender() above but also wrapped with 'react-router'
 * in order to test navigation flows.
 *
 * Returns the same object as reduxRender()
 */
export function routerRender(element, initialState, initialRoute = '/') {
  return reduxRender(
    <MemoryRouter initialEntries={[initialRoute]}>{element}</MemoryRouter>,
    initialState
  );
}

/*
 * Declaratively test complex form validations
 *
 * getter   : function that receives a data-testid and returns an HTML element
 * formId   : data-testid of the form element
 * formData : object of type { [data-testid]: any } to populate the form
 * errors   : object of type { [data-testid]: String } with expected errors
 */
export function testValidation(getter, formId, { formData, errors }) {
  // Populate the specified form fields
  Object.keys(formData).forEach((fieldId, i) => {
    const field = getter(fieldId);
    field.value = formData[fieldId];
    Simulate.change(field);
  });

  // Submit the form
  Simulate.submit(getter(formId));

  // Check expected errors
  Object.keys(errors).forEach((fieldId, i) => {
    expect(getter(`${fieldId}-error`).textContent).toEqual(
      expect.stringContaining(errors[fieldId])
    );
  });
}

/*
 * Add dataset property (data-foo attribute values) to nodes
 *
 * This hack is required because jsdom has no support for dataset yet
 * which prevents us from testing components that rely on it.
 * See https://github.com/jsdom/jsdom/issues/961
 *
 * element   : DOM node to add the dataset to
 * dataAttrs : attribute names to add to the node
 *
 * Usage:
 * <button data-testid="btn" data-modal="buy" data-foo="bar">Click</button>
 *
 * const btn = withDataset(getByTestId('btn'), 'modal', 'foo')
 * // btn.dataset === { modal: 'buy', foo: 'bar'}
 *
 */
export function withDataset(element, ...dataAttrs) {
  element.dataset = dataAttrs.reduce((acc, attr) => {
    acc[attr] = element.getAttribute(`data-${attr}`);
    return acc;
  }, {});
  return element;
}

/**
 * Return UTC timestamps relative to the current time
 * Useful for setting up different auction scenarios.
 */
export const twoWeeksAgo = () => Date.now() / 1000 - 60 * 60 * 24 * 7 * 2;
export const oneHourAgo = () => Date.now() / 1000 - 60 * 60;
export const inOneHour = () => Date.now() / 1000 + 60 * 60;
export const inOneWeek = () => Date.now() / 1000 + 60 * 60 * 24 * 7;

/*
 * Returns a common initial state for Redux store that is useful for most tests
 * Accepts an optional object with overrides that will be deeply merged with
 * the base state object.
 */
export function getInitialState(overrides = {}) {
  const baseState = {
    config,
    connectivity: { isOnline: true },
    blockchain: { height: 1, gasPrice: config.DEFAULT_GAS_PRICE },
    converter: {
      status: {
        availableEth: '100',
        availableLmr: '100',
        currentPrice: '10'
      }
    },
    auction: {
      status: {
        nextAuctionStartTime: inOneHour(),
        tokenRemaining: '1',
        currentAuction: 10,
        currentPrice: '33000000000',
        genesisTime: twoWeeksAgo()
      }
    },
    session: { isLoggedIn: true },
    rates: { ETH: { token: 'ETH', price: 1 } },
    wallets: {
      syncStatus: 'up-to-date',
      active: 'foo',
      allIds: ['foo'],
      byId: {
        foo: {
          addresses: {
            '0x15dd2028C976beaA6668E286b496A518F457b5Cf': {
              token: {
                [config.LMR_TOKEN_ADDR]: { balance: '5000000000000000000000' }
              },
              balance: '5000000000000000000000',
              transactions: []
            }
          }
        }
      }
    }
  };
  return merge({}, baseState, overrides);
}

/*
 * A sample transaction object. Useful for populating state in tests.
 */
export function getDummyTransaction() {
  return {
    meta: {
      lumerin: {},
      tokens: {
        [config.LMR_TOKEN_ADDR]: {
          event: 'Transfer',
          processing: false,
          value:
            '0x000000000000000000000000000000000000000000000001e5b8fa8fe2ac0000'
        }
      }
    },
    receipt: {
      blockHash: '0x1234567890abcdef',
      blockNumber: 407574,
      cumulativeGasUsed: 38484,
      gasUsed: 38484,
      transactionHash: '0x1234567890abcdef',
      transactionIndex: 0
    },
    transaction: {
      transactionIndex: 0,
      blockNumber: 407574,
      blockHash: '0x1234567890abcdef',
      hash: '0x1234567890abcdef',
      value: '0',
      gasPrice: '5000000000',
      gas: 76968,
      from: '0x15dd2028C976beaA6668E286b496A518F457b5Cf',
      to: '0xde806D6efD432CDeE42573760682D99eDEdC1d89'
    }
  };
}
