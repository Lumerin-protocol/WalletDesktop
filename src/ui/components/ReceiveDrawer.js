import PropTypes from 'prop-types';
import styled from 'styled-components';
import Drawer from './Drawer';
import React from 'react';

const Container = styled.div`
  padding: 0 4.8rem;
`;

export default class Receive extends React.Component {
  static propTypes = {
    onRequestClose: PropTypes.func.isRequired,
    isOpen: PropTypes.bool.isRequired
  };

  render() {
    const { onRequestClose, isOpen } = this.props;

    return (
      <Drawer onRequestClose={onRequestClose} isOpen={isOpen}>
        <Container>
          <h1>Receive</h1>
        </Container>
      </Drawer>
    );
  }
}
