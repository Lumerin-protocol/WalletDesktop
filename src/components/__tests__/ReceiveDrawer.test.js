import * as testUtils from '../../testUtils';
import ReceiveDrawer from '../dashboard/ReceiveDrawer';
import { Simulate } from 'react-testing-library';
import React from 'react';
import 'react-testing-library/extend-expect';

const initialState = testUtils.getInitialState();

const element = <ReceiveDrawer onRequestClose={jest.fn()} isOpen />;

describe('<ReceiveDrawer/>', () => {
  it('matches its snapshot', () => {
    const { container } = testUtils.reduxRender(element, initialState);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the active address', () => {
    const { getByTestId } = testUtils.reduxRender(element, initialState);
    const address = Object.keys(initialState.wallet.address);
    expect(getByTestId('address')).toHaveTextContent(address);
  });

  it.skip('displays a message when the Copy button is clicked and the address copied', async () => {
    const { getByTestId, queryByTestId } = testUtils.reduxRender(
      element,
      initialState
    );
    expect(queryByTestId('btn-label')).toHaveTextContent('Copy');
    Simulate.click(getByTestId('copy-btn'));
    expect(queryByTestId('btn-label')).toHaveTextContent(
      'Copied to clipboard!'
    );
  });
});
