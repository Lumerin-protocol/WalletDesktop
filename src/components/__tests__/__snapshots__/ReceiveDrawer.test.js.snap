// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReceiveDrawer/> matches its snapshot 1`] = `
<header
  class="sc-eHgmQL faXzUG"
>
  <h1
    class="sc-cvbbAY ewAxGX"
    data-testid="drawer-title"
  >
    Receive Transaction
  </h1>
  <button
    class="sc-jWBwVP litTlT"
    data-testid="drawer-close-btn"
  >
    <svg
      class="sc-bxivhb hXQRLM"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 24C5.386 24 0 18.614 0 12S5.386 0 12 0s12 5.386 12 12-5.386 12-12 12zM12 .96C5.914.96.96 5.914.96 12c0 6.086 4.954 11.04 11.04 11.04 6.086 0 11.04-4.954 11.04-11.04C23.04 5.914 18.086.96 12 .96zm3.984 15.696L12 12.672l-3.984 3.984-.682-.682 3.984-3.984-3.984-3.984.682-.681L12 11.309l3.984-3.984.682.681-3.984 3.984 3.984 3.984-.682.682z"
      />
    </svg>
  </button>
</header>
`;
